lib/react_client/react_interop.dart

```diff
@@ -127,15 +130,171 @@ class JsRef {
+///   // We can pass it along to LogProps as a regular prop, e.g. "forwardedRef"
+///   // And it can then be attached to the Component.
+///   return _logPropsHoc({...props, 'forwardedRef': ref});
+///   // Optional: Make the displayName more useful for the React dev tools.
+///   // See: https://reactjs.org/docs/forwarding-refs.html#displaying-a-custom-name-in-devtools
+/// }, displayName: 'LogProps(${_logPropsHoc.type.displayName})');
+///
+/// // ---------- Component Consumption ----------
+/// void main() {
+///   setClientConfiguration();
+///   final ref = react.createRef<Element>();
+///
+///   react_dom.render(LogProps({'component': FancyButton, 'ref': ref}),
+///       querySelector('#idOfSomeNodeInTheDom'));
+///
+///   // You can still get a ref directly to the DOM button:
+///   final buttonNode = ref.current;
+/// }
+/// ```
 /// See: <https://reactjs.org/docs/forwarding-refs.html>.
-ReactJsComponentFactoryProxy forwardRef(Function(Map props, Ref ref) wrapperFunction) {
-  var hoc = React.forwardRef(allowInterop((JsMap props, JsRef ref) {
+ReactJsComponentFactoryProxy forwardRef(
+  Function(Map props, Ref ref) wrapperFunction, {
+  String displayName = 'Anonymous',
+}) {
+  final wrappedComponent = allowInterop((JsMap props, JsRef ref) {
     final dartProps = JsBackedMap.backedBy(props);
+    for (var value in dartProps.values) {
+      if (value is Function) {
+        // Tag functions that came straight from the JS
+        // so that we know to pass them through as-is during prop conversion.
+        isRawJsFunctionFromProps[value] = true;
+      }
+    }
+
     final dartRef = Ref.fromJs(ref);
     return wrapperFunction(dartProps, dartRef);
-  }));
+  });
+  defineProperty(wrappedComponent, 'displayName', jsify({'value': displayName}));
+
+  var hoc = React.forwardRef(wrappedComponent);
+
+  return ReactJsComponentFactoryProxy(hoc);
+}
+
+/// Shorthand convenience function to wrap a JS component that utilizes `forwardRef`.
+///
+/// __Example:__
+///
+/// ```dart
+/// // Assuming that your app includes the JS file containing the MaterialUI components...
+/// @JS()
+/// class MaterialUI {
+///   external static ReactClass get TextField;
+/// }
+///
+/// /// This makes it so that the `inputRef` set on the Dart `TextField` component
+/// /// gets passed all the way to the root <input> node rendered by the JS `MaterialUI.TextField` component.
+/// final TextField = forwardRefToJsComponent(MaterialUI.IconButton, additionalRefPropKeys: ['inputRef']);
+///
+/// void main() {
+///   setClientConfiguration();
+///   final textFieldInputNodeRef = react.createRef<TextInputElement>();
+///
+///   react_dom.render(TextInput({'inputRef': textFieldInputNodeRef}), querySelector('#idOfSomeNodeInTheDom'));
+///
+///   // Prints the value of the <input> rendered by the JS `MaterialUI.TextField` component.
+///   print(textFieldInputNodeRef.current.value);
+/// }
+/// ```
+ReactJsComponentFactoryProxy forwardRefToJs(
```

Comment:
Taking a fresh look at what this functionality provides, I think this name is misleading, since it doesn't really provide similar functionality to forwardRef.

What about something like `withJsRefsConverted`? Also, `additionalRefPropKeys` should probably be renamed to not have `additional`, since this function doesn't really do anything without them

Reply:
I'm confused what you mean by _"doesn't really provide similar functionality to forwardRef"_.

This function returns the return value of a call to `forwardRef`.

Reply:
Right, but it doesn't allow you to forward a ref onto a different component/prop. It's really just a HOC that's using `forwardRef`.

For example, the last snippet in this section: https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components
> ```jsx
> function logProps(Component) {
>   class LogProps extends React.Component {
>     componentDidUpdate(prevProps) {
>       console.log('old props:', prevProps);
>       console.log('new props:', this.props);
>     }
> 
>     render() {
>       const {forwardedRef, ...rest} = this.props;
> 
>       // Assign the custom prop "forwardedRef" as a ref
>       return <Component ref={forwardedRef} {...rest} />;
>     }
>   }
> 
>   // Note the second param "ref" provided by React.forwardRef.
>   // We can pass it along to LogProps as a regular prop, e.g. "forwardedRef"
>   // And it can then be attached to the Component.
>   return React.forwardRef((props, ref) => {
>     return <LogProps {...props} forwardedRef={ref} />;
>   });
> }
> ```

`logProps` isn't `logPropsForwardRef`, it just uses it so that the HOC passes the ref through.

Reply:
@greglittlefield-wf I disagree that from an API clarity standpoint - `withJsRefsConverted` is more clear about the purpose of the function.

The purpose of the function is so that refs get forwarded.  All the HOC wrapping is semantics, IMO. The function does more than just convert refs... it also **forwards** them to the JS component.

Reply:
Okay, so taking a step back...

At this point, this function is just a wrapper around `ReactJsComponentFactoryProxy`, and the `forwardRef` call really serves no purpose. 

With the following changes:
```diff
-  return forwardRef((props, ref) {
-    return jsFactoryProxy({
-      ...props,
-      'ref': ref,
-    }, props['children']);
-    // Convert DOM props for the HOC so the consumer API callbacks that involve events receive Dart `SyntheticEvent`s.
-  }, displayName: '${displayName ?? jsClassComponent.displayName}');
+  return jsFactoryProxy;
```
The behavior would be exactly the same.

So, why do we need this function at all, if it's just a wrapper around `ReactJsComponentFactoryProxy`? Consumers should just use that instead.